Fix for users receiving a mustAuth response from the backend even though they set the password in this session.

shane-tomlinson opened the pull request:
If a user has set the password in this session (either dialog or main site) and they receive a mustAuth response from the verification poll, authenticate the user with the already entered email address and password.

This fixes users having to enter their password a second time on the /forgot and /signup pages after the verification is complete.

issue #1682

Commit:
Fix users who verify an email address but are not authenticated to the correct credentials.

* No longer set a user's authentication credentials to "password" if the user completes an address verification.  A user who is authed to assertion level and verifies on a second browser must still verify their password.
* When the polling wsapi completes, if the status is "complete" but the user's auth level is not "password", change the status to "mustAuth" which will force the user to enter their password if they have not done so already.

Commit:
Merge commit 'd02820f03b328053b61507a2ffa39e6a0b2acdb1' into issue_1682_verify_authentication

Merge of user.js and its unit tests.

Conflicts:
	resources/static/common/js/user.js
	resources/static/test/cases/common/js/user.js

Commit:
Update the comments.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -103,17 +105,26 @@ BrowserID.User = (function() {
+    // These are saved for the addressVerificationPoll.  If there is
+    // a stagedEmail or stagedPassword when the poll completes, try to
+    // authenticate the user.
+    stagedEmail = email;
```
Every time an email/password is staged, save off the staged email and password to use on completion.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -148,6 +159,68 @@ BrowserID.User = (function() {
 
   function addressVerificationPoll(checkFunc, email, onSuccess, onFailure) {
+    function userVerified(completionStatus) {
+      if (stagedEmail && stagedPassword) {
```
If there is a stagedEmail and stagedPassword, re-authenticate the user using the stagedEmail and stagedPassword.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -148,6 +159,68 @@ BrowserID.User = (function() {
+      // As soon as the registration comes back as complete, we should
+      // ensure that the stagedOnBehalfOf is cleared so there is no stale
+      // data.
+      storage.setReturnTo("");
```
This logic is taken verbatim from below.

Commit:
Merge remote-tracking branch 'origin/dev' into issue_1682_auth_after_verify

Conflicts:
	resources/static/common/js/user.js

shane-tomlinson commented:
Note that once this is merged, the old code in resources/static/dialog/js/modules/check_registration can be removed.

@lloyd - can you have a look at this?

seanmonstar commented:
@shane-tomlinson since the line comments seem needed to explain what's going on, should they be in the code instead?

shane-tomlinson commented:
@seanmonstar - outside of the last comment about "this logic is taken verbatim from below", the comments are already in the code as well, and slightly more verbose. I wanted to draw attention to them in the pull request.

Commit:
Merge branch 'dev' into issue_1682_auth_after_verify

Commit:
Cleanup. Removing the post-verification authentication/email sync from check_registration.

All of this is now done in user.js so that the logic can be shared between the dialog and main site.

seanmonstar commented:
This looks fine to me, the only thing stopping me from clicking merge is that it's 3 months old. I can click?

Commit:
Merge branch 'dev' into issue_1682_auth_after_verify
